### PR TITLE
Updates to read_csv for DateTimeIndex, and new tests

### PR DESCRIPTION
**Pull request by caseyclements**
Small changes while playing with our 'wide' financial use case.

### CODE REVIEW

**Comment by caseyclements** [comment]
The ci builds fail because I left a failing test there to draw attention to the case I need help on.

**Comment by caseyclements** [comment]
Although I didn't break the from_pandas test that I added, the dates in my example are descending, which pandas keeps as is, while dask resorts ascending.

File: dask/dataframe/io.py [hunk 1]
```diff
@@ -261,7 +261,7 @@ def categories_and_quantiles(fn, args, kwargs, index=None, categorize=None,
     if infer_header(fn, **kwargs):
         kwargs['header'] = 0
 
-    one_chunk = pd.read_csv(fn, *args, nrows=100, **kwargs)
+    one_chunk = pd.read_csv(fn, *args, **merge(kwargs, dict(nrows=100)))
```
Comment (mrocklin) [inline dask/dataframe/io.py]: Try `assoc(kwargs, 'nrows', 100)`

Comment (mrocklin) [inline dask/dataframe/io.py]: Grab assoc from toolz if not already around.

File: dask/dataframe/io.py [hunk 2]
```diff
@@ -271,7 +271,7 @@ def categories_and_quantiles(fn, args, kwargs, index=None, categorize=None,
         category_columns = []
     cols = category_columns + ([index] if index else [])
 
-    dtypes = dict((c, one_chunk.dtypes[c]) for c in cols)
+    dtypes = dict((c, one_chunk.dtypes[c]) for c in cols if c not in kwargs.get('parse_dates', ()))
```
Comment (mrocklin) [inline dask/dataframe/io.py]: Wrap this around to the next line.  We try to keep to 80 character lines.

Comment (mrocklin) [inline dask/dataframe/io.py]: This is a soft rule, we break it sometimes, but it's nice to do it when easy.

**Comment by caseyclements** [comment]
With assoc, will it override nrows if it already exists?

File: dask/dataframe/tests/test_io.py [hunk 3]
```diff
@@ -46,6 +46,47 @@ def test_read_csv():
...
+2015-08-28,198.50,199.839996,197.919998,199.240005,143298900,199.240005
+2015-08-27,197.020004,199.419998,195.210007,199.160004,266244700,199.160004
+2015-08-26,192.080002,194.789993,188.369995,194.679993,328058100,194.679993
+2015-08-25,195.429993,195.449997,186.919998,187.229996,353966700,187.229996
+2015-08-24,197.630005,197.630005,182.399994,189.550003,478672400,189.550003
+2015-08-21,201.729996,203.940002,197.520004,197.630005,328271500,197.630005
+2015-08-20,206.509995,208.289993,203.899994,204.009995,185865600,204.009995
+2015-08-19,209.089996,210.009995,207.350006,208.279999,167316300,208.279999
+2015-08-18,210.259995,210.679993,209.699997,209.929993,70043800,209.929993
+""".strip()
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Just checking, but these values are either random or publicly available, yes?

File: dask/dataframe/tests/test_io.py [hunk 4]
```diff
@@ -46,6 +46,47 @@ def test_read_csv():
...
+
+def test_read_csv_with_datetime_index_partitions_one():
+    with filetext(timeseries) as fn:
+        df = pd.read_csv(fn, index_col=0, header=0, usecols=[0, 4], parse_dates=['Date'])
+        # nrows set to explicitly set to single chunk
+        ddf = dd.read_csv(fn, index='Date', header=0, usecols=[0, 4], parse_dates=['Date'],  nrows=1000)
+        np.all(df == ddf.compute())
+        # because fn is so small, by default, this will only be one chunk
+        ddf = dd.read_csv(fn, index='Date', header=0, usecols=[0, 4], parse_dates=['Date'])
+        np.all(df == ddf.compute())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Two things, first, this `np.all` doesn't raise an assertion if it's false, it just returns false, so this line doesn't have an impact.  You probably wanted the following

    assert np.all(df == ddf.compute())

However, we actually have a special function for this (which includes the assert.)  Try the following instead:

    eq(df, ddf)

File: dask/dataframe/tests/test_io.py [hunk 5]
```diff
@@ -46,6 +46,47 @@ def test_read_csv():
...
+
+def test_read_csv_with_datetime_index_partitions_n():
+    with filetext(timeseries) as fn:
+        df = pd.read_csv(fn, index_col=0, header=0, usecols=[0, 4], parse_dates=['Date'])
+        # because fn is so small, by default, set chunksize small
+        ddf = dd.read_csv(fn, index='Date', header=0, usecols=[0, 4], parse_dates=['Date'], chunkbytes=400)
+        print(df)
+        print(ddf.compute())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: We should remove the print statements

File: dask/dataframe/tests/test_io.py (hunk 4 above)
Comment (caseyclements) [inline dask/dataframe/tests/test_io.py]: Nice. That was a typo. The assert should have been there.

Comment (caseyclements) [inline dask/dataframe/tests/test_io.py]: And yes, these are downloaded from yahoo finance.

File: dask/dataframe/tests/test_io.py [hunk 6]
```diff
@@ -46,6 +46,47 @@ def test_read_csv():
...
+        np.all(df == ddf.compute())
+
+def test_from_pandas_with_datetime_index():
+    with filetext(timeseries) as fn:
+        df = pd.read_csv(fn, index_col=0, header=0, usecols=[0, 4], parse_dates=['Date'])
+        ddf = dd.from_pandas(df, 2)
+        assert(df.index.dtype == ddf.index.compute().dtype)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: The dtype checks will happen in the `eq` function if you use it.  Actually, these use `pandas.utils.testing.assert_frame_equal` which checks *lots* of things.

File: dask/dataframe/tests/test_io.py [hunk 7]
```diff
@@ -712,3 +753,10 @@ def test_hdf_globbing():
         tm.assert_frame_equal(res.compute(), pd.concat([df] *  3))
     finally:
         shutil.rmtree(tdir)
+
+
+if __name__ == "__main__":
+    test_read_csv_with_datetime_index_partitions_one()
+    test_read_csv_with_datetime_index_partitions_n()
+    test_from_pandas_with_datetime_index()
+    test_usecols()
```
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: We should avoid these.  Instead I run tests by the following

    mrocklin@notebook:~/workspace/dask$ py.test dask/dataframe/tests/test_io.py

File: .gitignore [hunk 8]
```diff
@@ -7,3 +7,4 @@ dist/
 log.*
 log
 .coverage
+.DS_Store
```
Comment (mrocklin) [inline .gitignore]: Just out of curiosity, to what application does this belong?

**Comment by mrocklin** [comment]
The approach taken here seems useful to me.  I've left some nit-picky style changes.  I can check out the failing test itself sometime tomorrow.

File: dask/dataframe/tests/test_io.py (hunk 7 above)
Comment (caseyclements) [inline dask/dataframe/tests/test_io.py]: Yeah - I run tests that way, too, but for debugging purposes, with an IDE like pyCharm. That one doesn't break when running py.test. From bash, it's set_trace or nothing, I guess. it's convenient. I will remove if you wish.

File: dask/dataframe/tests/test_io.py (hunk 4 above)
Comment (caseyclements) [inline dask/dataframe/tests/test_io.py]: In what module is eq()?

File: .gitignore (hunk 8 above)
Comment (caseyclements) [inline .gitignore]: OSX Finder puts them in every folder it touches. It's terrible.

File: dask/dataframe/tests/test_io.py (hunk 4 above)
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: There is an `eq` function in this testing file.

File: dask/dataframe/tests/test_io.py (hunk 7 above)
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Removing would be good.

**Comment by mrocklin** [comment]
When you add new commits to your branch on github this PR will update automatically

**Comment by mrocklin** [comment]
This commit might be of use https://github.com/mrocklin/dask/commit/fda4e03d909c615369cf1b2b8a7900c16a3df5f8

You can cherry pick it into your branch with the following commands

```
git remote add mrocklin git@github.com:mrocklin/dask.git
git remote update mrocklin
git cherry-pick fda4e03d909c615369cf1b2b8a7900c16a3df5f8
```

**Comment by mrocklin** [comment]
@caseyclements if you're ok handling the style issues I'm happy handling the datetime index issues.

**Comment by mrocklin** [comment]
These tests are really useful

**Comment by caseyclements** [comment]
I'm fine updating the style issues. Might possibly have to be done tomorrow. I am committed to finishing a couple client tasks tonight.

**Comment by mrocklin** [comment]
OK.  Let me know if there is anything that I can do to help here.

**Comment by mrocklin** [comment]
Closing in favor of https://github.com/blaze/dask/pull/737

**Pull request closed without merging**